### PR TITLE
WebapiAsync Swagger

### PR DESCRIPTION
**Pull request by careys7**
### Description

Have migrated old commits to new PR pointed to CE, and merged the EE and CE code changes together into one branch. 

Basic integration is working but unable to finish acceptance tests due to outstanding issues in `async-webapi-2`.

**To be completed:**

- [ ] Add nicer Swagger UI paths. Currently like this https://bulk-api/swagger/?type=async
- [x] Finalise actual response description, HTTP response code
- [ ] Add unit tests for `ServiceMetadata` plugin 
- [ ] Code formatting 
- [x] Integration test for schema output 
- [x] Implement final schema URL in tests, blocks (eg `/rest/async/schema/`, `/async/schema` etc)

### Fixed Issues (if relevant)

1. magento-engcom/bulk-api#11 Change Swagger generator

### Manual testing scenarios
1. Visit https://bulk/swagger/?type=async in browser, expect to see async
2. Visit https://bulk/swagger/ in browser, expect to see rest 

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/WebapiAsync/etc/di.xml [hunk 1]
```diff
@@ -42,4 +42,8 @@
         </arguments>
     </type>
 
+    <type name="Magento\Webapi\Model\ServiceMetadata">
+        <plugin name="webapiServiceMetadataAsync" type="Magento\WebapiAsync\Plugin\ServiceMetadata" />
+    </type>
+
```
Comment (careys7) [inline app/code/Magento/WebapiAsync/etc/di.xml]: @vrann is this best under `webapi_async/di.xml`?

File: app/code/Magento/WebapiAsync/Model/ServiceConfig.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+     * Initialize dependencies.
+     *
+     * @param WebapiCache $cache
+     * @param Reader $configReader
+     * @param SerializerInterface|null $serializer
+     */
+    public function __construct(
+        WebapiCache $cache,
+        Reader $configReader,
+        SerializerInterface $serializer = null
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig.php]: Please inject this argument as required and do not use object manager in the constructor.
All this is necessary only for backward compatibility of existing classes and does not make sense for new classes.

File: app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * @param \DOMElement $serviceNode
+     * @return bool
+     */
+    private function getSynchronousMethodInvocationOnly(\DOMElement $serviceNode)
+    {
+        $synchronousInvocationOnlyNodes = $serviceNode->getElementsByTagName('synchronousInvocationOnly');
+
+        // This should be caught by the xsd, but covers some un-expected scenario
+        if (count($synchronousInvocationOnlyNodes) > 1) {
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php]: We usually do not duplicate validation performed by XSD. XSD validation is enabled in developer mode and any errors should be caught during development or during tests execution.

Keeping duplicate validation leads to the same issues as any other code duplication. Please consider removing duplicate validation for all existing XSD rules.

File: app/code/Magento/WebapiAsync/Model/ServiceConfig/SchemaLocator.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * Web API config schema locator.
+ */
+class SchemaLocator implements \Magento\Framework\Config\SchemaLocatorInterface
+{
+    /**
+     * Path to corresponding XSD file with validation rules for merged config
+     *
+     * @var string
+     */
+    protected $_schema = null;
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig/SchemaLocator.php]: Please do not use protected fields and underscores in names.

File: app/code/Magento/WebapiAsync/Model/ServiceConfig/SchemaLocator.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\WebapiAsync\Model\ServiceConfig;
+
+use Magento\Framework\Module\Dir;
+
+/**
+ * Web API config schema locator.
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig/SchemaLocator.php]: Please fix description

File: app/code/Magento/WebapiAsync/Test/Unit/Model/ServiceConfig/ConverterTest.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * See COPYING.txt for license details.
+ */
+namespace Magento\WebapiAsync\Test\Unit\Model\ServiceConfig;
+
+class ConverterTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var \Magento\WebapiAsync\Model\ServiceConfig\Converter
+     */
+    protected $_model;
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Test/Unit/Model/ServiceConfig/ConverterTest.php]: Please do not use `protected` and underscores. Check all new classes across the PR.

**Comment by careys7** [comment]
Swagger UI should now render async schema at the following:

https://bulk/swagger/?type=async

File: dev/tests/integration/testsuite/Magento/Webapi/Controller/Rest/SchemaRequestProcessorTest.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        // Check that an HTTP 200 response status is visible in the schema.
+        $this->assertRegExp('/200 Accepted/', $schema);
+    }
+
+    /**
+     * @return array
+     */
+    public function schemaRequestProvider()
+    {
+        return [
```
Comment (careys7) [inline dev/tests/integration/testsuite/Magento/Webapi/Controller/Rest/SchemaRequestProcessorTest.php]: @vrann can you confirm that these are all as expected? 

I get the below output when testing manually with any of the rest ones (`async` ones work fine)

```
<response>
<message>
Warning: preg_match(): Unknown modifier 'V' in /var/www/current/src/app/code/Magento/WebapiAsync/Controller/Rest/AsynchronousRequestProcessor.php on line 123
</message>
<trace>
#0 [internal function]: Magento\Framework\App\ErrorHandler->handler(2, 'preg_match(): U...', '/var/www/curren...', 123, Array) #1 /var/www/current/src/app/code/Magento/WebapiAsync/Controller/Rest/AsynchronousRequestProcessor.php(123): preg_match('/async/V\\d+/', 'schema') #2 /var/www/current/src/app/code/Magento/Webapi/Controller/Rest/RequestProcessorPool.php(39): Magento\WebapiAsync\Controller\Rest\AsynchronousRequestProcessor->canProcess(Object(Magento\Framework\Webapi\Rest\Request\Proxy)) #3 /var/www/current/src/app/code/Magento/Webapi/Controller/Rest.php(187): Magento\Webapi\Controller\Rest\RequestProcessorPool->getProcessor(Object(Magento\Framework\Webapi\Rest\Request\Proxy)) #4 /var/www/current/src/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http)) #5 /var/www/current/src/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent('dispatch', Array) #6 /var/www/current/src/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http)) #7 /var/www/current/src/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(26): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array) #8 /var/www/current/src/lib/internal/Magento/Framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http)) #9 /var/www/current/src/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch() #10 /var/www/current/src/lib/internal/Magento/Framework/App/Bootstrap.php(256): Magento\Framework\App\Http\Interceptor->launch() #11 /var/www/current/src/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor)) #12 {main}
</trace>
</response>
```

File: app/code/Magento/WebapiAsync/etc/di.xml (hunk 1 above)
Comment (vrann) [reply app/code/Magento/WebapiAsync/etc/di.xml]: we haven't introduced new area for async so it should stay here. Does it affect the ServiceMetadata for the REST routes?

**Pull request merged**